### PR TITLE
Fix gcc armel/arm64 build

### PR DESCRIPTION
**Pull request by gbalykov**
Mostly signed/unsigned comparisons, etc.

cc @alpencolt

### CODE REVIEW

File: src/coreclr/jit/codegenarm.cpp [hunk 1]
```diff
@@ -460,7 +460,7 @@ void CodeGen::genLclHeap(GenTree* tree)
         }
 
         // regCnt will be the total number of bytes to locAlloc
-        genSetRegToIcon(regCnt, amount, ((int)amount == amount) ? TYP_INT : TYP_LONG);
+        genSetRegToIcon(regCnt, amount, ((int)amount >= 0) ? TYP_INT : TYP_LONG);
```
Comment (sandreenko) [inline src/coreclr/jit/codegenarm.cpp:463]: this does not look right, a narrowing cast is implementation-defined in C++.
also, it looks like we don't support TYP_LONG in https://github.com/dotnet/runtime/blob/eee0ed206b9277f3ade51af3ceb3aadb6ff609c0/src/coreclr/jit/codegenarm.cpp#L183
that we are calling here, maybe use `ClrSafeInt ` and cast to int always? cc @echesakovMSFT

File: src/coreclr/pal/src/libunwind/src/arm/Gtrace.c [hunk 2]
```diff
@@ -526,7 +526,7 @@ tdep_trace (unw_cursor_t *cursor, void **buffer, int *size)
       break;
 
     case UNW_ARM_FRAME_SYSCALL:
-      printf("XXX1\n");
+      //printf("XXX1\n");
```
Comment (sandreenko) [inline src/coreclr/pal/src/libunwind/src/arm/Gtrace.c:529]: not familiar with this code, what was the problem?

**Comment by sandreenko** [comment]
cc @dotnet/jit-contrib

File: src/coreclr/pal/src/libunwind/src/arm/Gtrace.c (hunk 2 above)
Comment (gbalykov) [reply src/coreclr/pal/src/libunwind/src/arm/Gtrace.c:529]:
```
[   71s]   In file included from /home/abuild/rpmbuild/BUILD/coreclr-6.0.0/src/coreclr/pal/src/libunwind/src/arm/Ltrace.c:4:
[   71s]   /home/abuild/rpmbuild/BUILD/coreclr-6.0.0/src/coreclr/pal/src/libunwind/src/arm/Gtrace.c: In function '_ULarm_tdep_trace':
[   71s]   /home/abuild/rpmbuild/BUILD/coreclr-6.0.0/src/coreclr/pal/src/libunwind/src/arm/Gtrace.c:529:7: error: incompatible implicit declaration of built-in function 'printf' [-Werror]
[   71s]     529 |       printf("XXX1\n");
[   71s]         |       ^~~~~~
[   71s]   /home/abuild/rpmbuild/BUILD/coreclr-6.0.0/src/coreclr/pal/src/libunwind/src/arm/Gtrace.c:31:1: note: include '<stdio.h>' or provide a declaration of 'printf'
[   71s]      30 | #include <limits.h>
[   71s]     +++ |+#include <stdio.h>
[   71s]      31 |
```

File: src/coreclr/jit/codegenarm.cpp (hunk 1 above)
Comment (gbalykov) [reply src/coreclr/jit/codegenarm.cpp:463]: Yes, my bad, thanks. But it seems that original code is incorrect too

File: src/coreclr/ToolBox/superpmi/superpmi-shared/lightweightmap.h [hunk 3]
```diff
@@ -319,7 +319,7 @@ class LightWeightMap : public LightWeightMapBuffer
 
         // If we have RTTI, we can make this assert report the correct type. No RTTI, though, when
         // built with .NET Core, especially when built against the PAL.
-        AssertCodeMsg((ptr - bytes) == size, EXCEPTIONCODE_LWM, "%s - Ended with unexpected sizes %p != %x",
+        AssertCodeMsg((unsigned int)(ptr - bytes) == size, EXCEPTIONCODE_LWM, "%s - Ended with unexpected sizes %p != %x",
```
Comment (echesakovMSFT) [inline src/coreclr/ToolBox/superpmi/superpmi-shared/lightweightmap.h:322]: I would rewrite it as follows to avoid casting
```suggestion
        AssertCodeMsg(ptr == (bytes + size), EXCEPTIONCODE_LWM, "%s - Ended with unexpected sizes %p != %x",
```

File: src/coreclr/ToolBox/superpmi/superpmi-shared/lightweightmap.h [hunk 4]
```diff
@@ -658,7 +658,7 @@ class DenseLightWeightMap : public LightWeightMapBuffer
             ptr += bufferLength * sizeof(unsigned char);
         }
 
-        AssertCodeMsg((ptr - bytes) == size, EXCEPTIONCODE_LWM, "Ended with unexpected sizes %Ix != %x", ptr - bytes,
+        AssertCodeMsg((unsigned int)(ptr - bytes) == size, EXCEPTIONCODE_LWM, "Ended with unexpected sizes %Ix != %x", ptr - bytes,
```
Comment (echesakovMSFT) [inline src/coreclr/ToolBox/superpmi/superpmi-shared/lightweightmap.h:661]: Same here
```suggestion
        AssertCodeMsg(ptr == (bytes + size), EXCEPTIONCODE_LWM, "Ended with unexpected sizes %Ix != %x", ptr - bytes,
```

File: src/coreclr/jit/codegenarm.cpp (hunk 1 above)
Comment (echesakovMSFT) [reply src/coreclr/jit/codegenarm.cpp:463]: I don't think this is relevant to `CodeGen::instGen_Set_Reg_To_Reloc_Imm` issue in cross-bitness scenarious.

The change is for `localloc` (i.e. .NET way of `alloca`). 

According to ECMA-335 

*The `localloc` instruction allocates size (type `native unsigned int` or `U4`) bytes from the local
dynamic memory pool ...*

I agree with Sergey that we should always use `TYP_INT` here.
```suggestion
        genSetRegToIcon(regCnt, amount, TYP_INT);
```

File: src/coreclr/pal/src/CMakeLists.txt [hunk 5]
```diff
@@ -112,10 +112,6 @@ if(CLR_CMAKE_HOST_ARCH_ARM)
   endif()
 endif(CLR_CMAKE_HOST_ARCH_ARM)
 
-if (CMAKE_CXX_COMPILER_ID MATCHES "GNU")
```
Comment (echesakovMSFT) [inline src/coreclr/pal/src/CMakeLists.txt:115]: Can you please give an example why the change is needed? Is it assembler related?

File: src/coreclr/pal/src/libunwind/src/arm/Gtrace.c (hunk 2 above)
Comment (echesakovMSFT) [reply src/coreclr/pal/src/libunwind/src/arm/Gtrace.c:529]: Would adding `#include <stdio.h>` at the beginning of the file fix the issue or it would introduce some other errors? 
cc @janvorli

File: src/coreclr/pal/src/misc/jitsupport.cpp [hunk 6]
```diff
@@ -82,7 +82,7 @@ static const CpuCapability CpuCapabilities[] = {
 // If the capability name is not recognized or unused at present, zero is returned.
 static unsigned long LookupCpuCapabilityFlag(const char* start, size_t length)
 {
-    for (int i = 0; i < _countof(CpuCapabilities); i++)
+    for (int i = 0; i < (int)_countof(CpuCapabilities); i++)
```
Comment (echesakovMSFT) [inline src/coreclr/pal/src/misc/jitsupport.cpp:85]: I would avoid casting
```suggestion
    for (size_t i = 0; i < _countof(CpuCapabilities); i++)
```

File: src/coreclr/pal/src/CMakeLists.txt (hunk 5 above)
Comment (janvorli) [reply src/coreclr/pal/src/CMakeLists.txt:115]: This should stay, but the condition should be updated so that it gets added only for  Intel processors (CLR_CMAKE_HOST_ARCH_AMD64 or CLR_CMAKE_HOST_ARCH_I386).

Comment (gbalykov) [reply src/coreclr/pal/src/CMakeLists.txt:115]: arm assembler doesn't have this option, and linux x64 shows:
```
$ as --help|grep divide
  --divide                ignored
```

Why is this option needed?

Comment (janvorli) [reply src/coreclr/pal/src/CMakeLists.txt:115]: This was added for SunOS support. @am11 do you remember why was it needed?

Comment (am11) [reply src/coreclr/pal/src/CMakeLists.txt:115]: On SYS-V derivative platforms, `/` in assembly is treated as `#` rather than division. In context2.S, we are using division, so it breaks build on platforms, such as SunOS-likes and QNX.

Please note that GNU toolchain support is strictly added for community supported platforms in .NET, where LLVM toolchain is either not available or not as complete. If your platform  already has good support for LLVM toolchain, it is recommended (per its official support stauts) to use that instead.

Comment (gbalykov) [reply src/coreclr/pal/src/CMakeLists.txt:115]: Thanks for explanation. I've kept the original GNU condition and added x86/x64 check.

File: src/coreclr/pal/src/libunwind/src/arm/Gtrace.c (hunk 2 above)
Comment (gbalykov) [reply src/coreclr/pal/src/libunwind/src/arm/Gtrace.c]: I've removed this change from here

File: src/coreclr/jit/codegenarm.cpp (hunk 1 above)
Comment (gbalykov) [reply src/coreclr/jit/codegenarm.cpp]: Fixed, thanks

**Review by sandreenko — APPROVED** [review]
I see that the PR was updated, LGTM, thanks for making the changes.

**Review by echesakov — APPROVED** [review]

**Comment by BruceForstall** [comment]
Test failures are https://github.com/dotnet/runtime/issues/53311
